Feature Request to allow service to crash if direct memory is OOM

The direct memory of our service recently ran into an OOM exception for direct memory during testing. 

Our expectation initially was that the service would automatically restart because we enabled the jvm flag `-XX:+CrashOnOutOfMemoryError`. 

Later we found out that this flag is only used for heap or metaspace OOM. More details in this [openJDK bug repot](https://bugs.openjdk.org/browse/JDK-8155004)

Would it be possible to introduce a new flag that would react in the same way as `-XX:+CrashOnOutOfMemoryError` but includes direct memory.

root cause:
"Caused by: java.lang.OutOfMemoryError: Cannot reserve 16709 bytes of direct buffer memory (allocated: 10477782, limit: 10485760)",
"\tat java.base/java.nio.Bits.reserveMemory(Bits.java:178)",
"\tat java.base/java.nio.DirectByteBuffer.(DirectByteBuffer.java:121)",
"\tat java.base/java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:332)",
"\tat java.base/sun.nio.ch.Util.getTemporaryDirectBuffer(Util.java:243)",
"\tat java.base/sun.nio.ch.IOUtil.read(IOUtil.java:293)",
"\tat java.base/sun.nio.ch.IOUtil.read(IOUtil.java:266)",
"\tat java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.implRead(UnixAsynchronousSocketChannelImpl.java:525)",
"\t... 23 more"

Closed openjdk issue for similar request: https://github.com/openjdk/jdk/pull/16176#issuecomment-1760881555

--- a/src/hotspot/share/prims/unsafe.cpp
+++ b/src/hotspot/share/prims/unsafe.cpp
@@ -479,6 +479,13 @@ UNSAFE_LEAF (void, Unsafe_WriteBackPostSync0(JNIEnv *env, jobject unsafe)) {
   doWriteBackSync0(false);
 } UNSAFE_END
 
+// SapMachine 2025-02-05: Report error for DirectMemoryOom to exit the VM
+UNSAFE_ENTRY (void, Unsafe_ReportJavaOutOfMemory0(JNIEnv *env, jobject unsafe, jstring message)) {
+  ResourceMark rm;
+  char *utf_message = java_lang_String::as_utf8_string(JNIHandles::resolve_non_null(message));
+  report_java_out_of_memory(utf_message);
+} UNSAFE_END
+
 ////// Random queries
 
 static jlong find_field_offset(jclass clazz, jstring name, TRAPS) {
@@ -915,6 +922,8 @@ static JNINativeMethod jdk_internal_misc_Unsafe_methods[] = {
     {CC "shouldBeInitialized0", CC "(" CLS ")Z",         FN_PTR(Unsafe_ShouldBeInitialized0)},
 
     {CC "fullFence",          CC "()V",                  FN_PTR(Unsafe_FullFence)},
+    // SapMachine 2025-02-05: Report error for DirectMemoryOom to exit the VM
+    {CC "reportJavaOutOfMemory0", CC "(" LANG "String;)V", FN_PTR(Unsafe_ReportJavaOutOfMemory0)},
 };
 
 #undef CC

--- a/src/java.base/share/classes/java/nio/Bits.java
+++ b/src/java.base/share/classes/java/nio/Bits.java
@@ -68,6 +68,8 @@ class Bits {                            // package-private
     // -- Processor and memory-system properties --
 
     private static int PAGE_SIZE = -1;
+    // SapMachine 2025-02-05: Report error for DirectMemoryOom to exit the VM
+    private static boolean reportErrorOnDirectMemoryOom = Boolean.getBoolean("jdk.nio.reportErrorOnDirectMemoryOom");
 
     static int pageSize() {
         if (PAGE_SIZE == -1)
@@ -172,10 +174,15 @@ class Bits {                            // package-private
             }
 
             // no luck
-            throw new OutOfMemoryError
+            // SapMachine 2025-02-05: Report error for DirectMemoryOom to exit the VM
+            OutOfMemoryError error = new OutOfMemoryError
                 ("Cannot reserve "
                  + size + " bytes of direct buffer memory (allocated: "
                  + RESERVED_MEMORY.get() + ", limit: " + MAX_MEMORY +")");
+            if (reportErrorOnDirectMemoryOom) {
+                UNSAFE.reportJavaOutOfMemory0(error.getMessage());
+            }
+            throw error;
 
         } finally {
             if (interrupted) {

--- a/src/java.base/share/classes/jdk/internal/misc/Unsafe.java
+++ b/src/java.base/share/classes/jdk/internal/misc/Unsafe.java
@@ -3843,6 +3843,8 @@ public final class Unsafe {
     private native int arrayBaseOffset0(Class<?> arrayClass);
     private native int arrayIndexScale0(Class<?> arrayClass);
     private native int getLoadAverage0(double[] loadavg, int nelems);
+    // SapMachine 2025-02-05: Report error for DirectMemoryOom to exit the VM
+    public native void reportJavaOutOfMemory0(String message);
 
 
     /**

--- a/test/hotspot/jtreg/runtime/ErrorHandling/TestExitOnDirectOutOfMemoryError.java
+++ b/test/hotspot/jtreg/runtime/ErrorHandling/TestExitOnDirectOutOfMemoryError.java
@@ -1,0 +1,67 @@
+/*
+ * Copyright (c) 2015, 2025, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022 THL A29 Limited, a Tencent company. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test TestExitOnDirectOutOfMemoryError
+ * @summary Test using -XX:ExitOnOutOfMemoryError
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib
+ * @requires vm.flagless
+ * @run driver TestExitOnDirectOutOfMemoryError
+ */
+
+import java.nio.ByteBuffer;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+
+public class TestExitOnDirectOutOfMemoryError {
+
+    public static void main(String[] args) throws Exception {
+        if (args.length == 1) {
+            // This should guarantee to throw:
+            // java.lang.OutOfMemoryError: Cannot reserve 2147483647 bytes of direct buffer memory (allocated: 0, limit: 10485760)
+            try {
+                ByteBuffer byteBuffer = ByteBuffer.allocateDirect(Integer.MAX_VALUE);
+            } catch (OutOfMemoryError err) {
+                throw new Error("OOME didn't terminate JVM!");
+            }
+        }
+
+        // else this is the main test
+        ProcessBuilder pb = ProcessTools.createLimitedTestJavaProcessBuilder("-XX:+ExitOnOutOfMemoryError",
+                "-Xmx20m", "-XX:MaxDirectMemorySize=10m", "-Djdk.nio.reportErrorOnDirectMemoryOom=true",
+                TestExitOnDirectOutOfMemoryError.class.getName(), "throwOOME");
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+
+        /*
+         * Actual output should look like this:
+         * Terminating due to java.lang.OutOfMemoryError: Cannot reserve 2147483647 bytes of direct buffer memory (allocated: 0, limit: 10485760)
+         */
+        output.shouldHaveExitValue(3);
+        output.stdoutShouldNotBeEmpty();
+        output.shouldContain("Terminating due to java.lang.OutOfMemoryError: Cannot reserve 2147483647 bytes of direct" +
+                " buffer memory (allocated: 0, limit: 10485760)");
+        System.out.println("PASSED");
+    }
+}

